# Monstrinho 3.2 -
## Átomos não são bolinhas e ligações químicas não são pauzinhos” — Prof. Julio
####  Rômulo Emanuel Rabelo Cruz 24024

### Enunciado:

**Objetivo:** Utilize classes de Python para modelar elementos químicos e moléculas.

**Considerações do experimento:** Crie uma classe chamada Elemento. Todo elemento químico utilizado nesta tarefa deve ser uma instância desta classe. Toda instância
da classe Elemento deve ter os seguintes atributos: símbolo do elemento, número atômico
do elemento e peso atômico do elemento (fique à vontade para incluir outras informações
se quiser). Crie uma classe chamada Molecula. Esta classe deve receber um dicionário
onde as chaves representam os elementos e os valores representam a quantidade do respectivo elemento. A classe Molecula deve ter um método capaz de exibir o peso atômico da molécula criada e um método capaz gerar uma fórmula química (em string) para esta
molécula.

**Observação:** Não é necessário que a fórmula química siga as regras oficiais da química,
basta representar os elementos existentes e suas quantidades.


### Imports Necessários:

Importei uma tabela obtida no site 'https://pubchem.ncbi.nlm.nih.gov/' e pedi para o chat GPT me gerar um dataset de moléculas para testarmos o código

In [1]:
import pandas as pd

# Carregando csv da tabela periódica e csv de moléculas gerado pelo chat GPT
p_table = pd.read_csv('https://pubchem.ncbi.nlm.nih.gov/rest/pug/periodictable/CSV?response_type=display')

molecules = pd.read_csv('molecules_dataset.csv')

Moléculas:

In [2]:
molecules

Molecule Name  Hydrogen  Helium  Lithium  Beryllium  Boron  \
0                    Ethanol       6.0     0.0      0.0        0.0    0.0   
1                    Benzene       6.0     0.0      0.0        0.0    0.0   
2                 Chloroform       1.0     0.0      0.0        0.0    0.0   
3       Uranium Hexafluoride       0.0     0.0      0.0        0.0    0.0   
4       Phosphotungstic Acid       3.0     0.0      0.0        0.0    0.0   
5            Ferric Chloride       0.0     0.0      0.0        0.0    0.0   
6              Sulfuric Acid       2.0     0.0      0.0        0.0    0.0   
7                    Ammonia       3.0     0.0      0.0        0.0    0.0   
8            Sodium Chloride       0.0     0.0      0.0        0.0    0.0   
9          Calcium Carbonate       0.0     0.0      0.0        0.0    0.0   
10                   Methane       4.0     0.0      0.0        0.0    0.0   
11               Acetic Acid       4.0     0.0      0.0        0.0    0.0   
12          Titanium Dioxide       0.0     0.0      0.0        0.0    0.0   
13      Potassium Dichromate       0.0     0.0      0.0        0.0    0.0   
14                  Caffeine      10.0     0.0      0.0        0.0    0.0   
15   Hemoglobin (simplified)    1064.0     0.0      0.0        0.0    0.0   
16               Vitamin B12      88.0     0.0      0.0        0.0    0.0   
17  DNA Nucleotide (Adenine)      13.0     0.0      0.0        0.0    0.0   
18             Fullerene C60       0.0     0.0      0.0        0.0    0.0   
19           Plutonium Oxide       0.0     0.0      0.0        0.0    0.0   

    Carbon  Nitrogen  Oxygen  Fluorine  ...  Meitnerium  Darmstadtium  \
0      2.0       0.0     1.0       0.0  ...         0.0           0.0   
1      6.0       0.0     0.0       0.0  ...         0.0           0.0   
2      1.0       0.0     0.0       0.0  ...         0.0           0.0   
3      0.0       0.0     0.0       6.0  ...         0.0           0.0   
4      0.0       0.0    40.0       0.0  ...         0.0           0.0   
5      0.0       0.0     0.0       0.0  ...         0.0           0.0   
6      0.0       0.0     4.0       0.0  ...         0.0           0.0   
7      0.0       1.0     0.0       0.0  ...         0.0           0.0   
8      0.0       0.0     0.0       0.0  ...         0.0           0.0   
9      1.0       0.0     3.0       0.0  ...         0.0           0.0   
10     1.0       0.0     0.0       0.0  ...         0.0           0.0   
11     2.0       0.0     2.0       0.0  ...         0.0           0.0   
12     0.0       0.0     2.0       0.0  ...         0.0           0.0   
13     0.0       0.0     7.0       0.0  ...         0.0           0.0   
14     8.0       4.0     2.0       0.0  ...         0.0           0.0   
15  6800.0     174.0   180.0       0.0  ...         0.0           0.0   
16    63.0      14.0    14.0       0.0  ...         0.0           0.0   
17    10.0       5.0     4.0       0.0  ...         0.0           0.0   
18    60.0       0.0     0.0       0.0  ...         0.0           0.0   
19     0.0       0.0     2.0       0.0  ...         0.0           0.0   

    Roentgenium  Copernicium  Nihonium  Flerovium  Moscovium  Livermorium  \
0           0.0          0.0       0.0        0.0        0.0          0.0   
1           0.0          0.0       0.0        0.0        0.0          0.0   
2           0.0          0.0       0.0        0.0        0.0          0.0   
3           0.0          0.0       0.0        0.0        0.0          0.0   
4           0.0          0.0       0.0        0.0        0.0          0.0   
5           0.0          0.0       0.0        0.0        0.0          0.0   
6           0.0          0.0       0.0        0.0        0.0          0.0   
7           0.0          0.0       0.0        0.0        0.0          0.0   
8           0.0          0.0       0.0        0.0        0.0          0.0   
9           0.0          0.0       0.0        0.0        0.0          0.0   
10          0.0 

Tabela Periódica:

In [3]:
p_table

AtomicNumber Symbol         Name  AtomicMass CPKHexColor  \
0               1      H     Hydrogen    1.008000      FFFFFF   
1               2     He       Helium    4.002600      D9FFFF   
2               3     Li      Lithium    7.000000      CC80FF   
3               4     Be    Beryllium    9.012183      C2FF00   
4               5      B        Boron   10.810000      FFB5B5   
..            ...    ...          ...         ...         ...   
113           114     Fl    Flerovium  290.192000         NaN   
114           115     Mc    Moscovium  290.196000         NaN   
115           116     Lv  Livermorium  293.205000         NaN   
116           117     Ts   Tennessine  294.211000         NaN   
117           118     Og    Oganesson  295.216000         NaN   

                 ElectronConfiguration  Electronegativity  AtomicRadius  \
0                                  1s1               2.20         120.0   
1                                  1s2                NaN         140.0   
2                              [He]2s1               0.98         182.0   
3                              [He]2s2               1.57         153.0   
4                          [He]2s2 2p1               2.04         192.0   
..                                 ...                ...           ...   
113  [Rn]7s2 7p2 5f14 6d10 (predicted)                NaN           NaN   
114  [Rn]7s2 7p3 5f14 6d10 (predicted)                NaN           NaN   
115  [Rn]7s2 7p4 5f14 6d10 (predicted)                NaN           NaN   
116  [Rn]7s2 7p5 5f14 6d10 (predicted)                NaN           NaN   
117  [Rn]7s2 7p6 5f14 6d10 (predicted)                NaN           NaN   

     IonizationEnergy  ElectronAffinity        OxidationStates  \
0              13.598             0.754                 +1, -1   
1              24.587               NaN                      0   
2               5.392             0.618                     +1   
3               9.323               NaN                     +2   
4               8.298             0.277                     +3   
..                ...               ...                    ...   
113               NaN               NaN           6, 4,2, 1, 0   
114               NaN               NaN                   3, 1   
115               NaN               NaN             +4, +2, -2   
116               NaN               NaN         +5, +3, +1, -1   
117               NaN               NaN  +6, +4, +2, +1, 0, -1   

              StandardState  MeltingPoint  BoilingPoint   Density  \
0                       Gas         13.81         20.28  0.000090   
1                       Gas          0.95          4.22  0.000179   
2                     Solid        453.65       1615.00  0.534000   
3                     Solid       1560.00       2744.00  1.850000   
4                     Solid       2348.00       4273.00  2.370000   
..                      ...           ...           ...       ...   
113  Expected to be a Solid           NaN           NaN       NaN   
114  Expected to be a Solid           NaN           NaN       NaN   
115  Expected to be a Solid           NaN           NaN       NaN   
116  Expected to be a Solid           NaN           NaN       NaN   
117    Expected to be a Gas           NaN           NaN       NaN   

                GroupBlock YearDiscovered  
0                 Nonmetal           1766  
1                Noble gas           1868  
2             Alkali metal           1817  
3     Alkaline earth metal           1798  
4                Metalloid           1808  
..                     ...            ...  
113  Post-transition metal           1998  
114  Post-transition metal           2003  
115  Post-transition metal           2000  
116                Halogen           2010  
117              Noble gas           2006  

[118 rows x 17 columns]

### Criando classes:

Criando a classe 'Element', na qual será necessário declarar o seu símbolo, número atômico e massa atômca. Também criei uma classe 'Molecule' que recebe um dicionário com o nome do elemento e suas propriedades.

In [4]:
# Criando a classe Element conforme o enunciado e armazendando os parêmetros inseridos como instâncias
class Element:
    def __init__(self,symbol:str,a_number:int,a_mass:float):
        self.symbol = symbol
        self.a_number = a_number
        self.a_mass = a_mass

In [5]:
# Criando a classe Molecule baseado no enunciado (com entrada um dicionário, que será armazenado na instância amount)
class Molecule:
    def __init__(self, dictionary:dict):
        self.amount = dictionary 
                
    def show_mass(self): # Função que mostra a massa molecular
        self.mass = 0
        for element in self.amount:
            self.mass += self.amount.get(element) * element.a_mass
        return self.mass

    def generate_formula(self): # Função de gerar fórmula, que pegará a chave(nome do elemento) e o seu valor respectivo(quantidade de átomos)
        self.formula = ''
        for element in self.amount:
            self.formula += element.symbol
            if self.amount.get(element) > 1:
                self.formula += str(self.amount.get(element))
        return self.formula

### Criando objetos utilizando as classes criadas:

Coloquei um 'for' para criar um objeto da classe 'Element' para cada elemento químico presente no 'DataFrame' da tabela periódica, iterando suas respectivas informações

In [6]:
# Criando um objeto Element para cada elemento da tabela periódica com base nos dados da tabela periódica
for line in range(0,len(p_table)):
    globals()[p_table['Name'][line]] = Element(p_table['Symbol'][line], p_table['AtomicNumber'][line], p_table['AtomicMass'][line])

Como a classe 'Molecule' pega um dicionário como parâmetro para criar o objeto molécula, será criado um dicionário para cada molécula presente no 'DataFrame' de moléculas. Depois disso, será criado um objeto de classe 'Molecule' para cada molécula presente no 'DataFrame' de moléculas utilizando o próprio dicionario homônimo obtido anteriormente, para que assim sobrescreva a variável, reduzindo o armazenamento desnecessário desses dicionários.

In [7]:
for molecule in range(len(molecules)):
    globals()[molecules.iloc[molecule, 0]] = {} # Cria dicionário vazio
    for column in range(1, len(molecules.columns)): # Estrutura de repetição que preencherá os dicionários
        if molecules.iloc[molecule, column] > 0: # Condição para preencher somente átomos presentes na molécula
            column_name =  globals().get(molecules.columns[column]) # Itera o nome do elemento e cria a chave com esse nome
            globals()[molecules.iloc[molecule, 0] ][column_name] = molecules.iloc[molecule, column].astype(int) # Itera o valor da quantidade de átomos presentes na molécula e dá o valor da chave
    globals()[molecules.iloc[molecule, 0]] = (Molecule(globals()[molecules.iloc[molecule, 0]])) # Sobreescreve o dicionário com um objeto 'Molecule'

Esse objeto será criado com o nome da molécula que está armazendado na coluna [0], a sua linha será representada por 'molecule', um inteiro que representa a linha em que este elemento está presente.

### Testes:

In [8]:
for molecule in range(len(molecules)):
    print(globals()[molecules.iloc[molecule, 0]].generate_formula(), globals()[molecules.iloc[molecule, 0]].show_mass())

H6C2O 46.069
H6C6 78.114
HCCl3 119.369
F6U 352.01931895999996
H3O40PW12 2880.037762
Cl3Fe 162.19
H2O4S 98.082
H3N 17.031
NaCl 58.4397693
CO3Ca 100.088
H4C 16.043
H4C2O2 60.052
O2Ti2 127.732
O7K2Cr2 294.1816
H10C8N4O2 194.194
H1064C6800N174O180S8Fe4 88544.26999999999
H88C63N14O14PCo 1355.387952
H13C10N5O4P 298.218762
C60 720.66
O2Pu 276.0622


#### Informações das moléculas:
Molecule Name | Molecular Formula | Molecular Mass (g/mol)

Ethanol | C₂H₆O | 46.07

Benzene | C₆H₆ | 78.11

Chloroform | CHCl₃ | 119.38

Uranium Hexafluoride | UF₆ | 352.02

Phosphotungstic Acid | H₃PW₁₂O₄₀ | 2880.2

Ferric Chloride | FeCl₃ | 162.2

Sulfuric Acid | H₂SO₄ | 98.08

Ammonia | NH₃ | 17.03

Sodium Chloride | NaCl | 58.44

Calcium Carbonate | CaCO₃ | 100.09

Methane | CH₄ | 16.04

Acetic Acid | C₂H₄O₂ | 60.05

Titanium Dioxide | TiO₂ | 79.87

Potassium Dichromate | K₂Cr₂O₇ | 294.18

Caffeine | C₈H₁₀N₄O₂ | 194.19

Hemoglobin (simplified) | C₆₈₀₀H₁₀₆₄N₁₇₄O₁₈₀S₈Fe₄ | ~64,500

Vitamin B12 | C₆₃H₈₈CoN₁₄O₁₄P | 1355.39

DNA Nucleotide (Adenine) | C₁₀H₁₃N₅O₄P | 331.2]

Fullerene C60 | C₆₀ | 720.66

Plutonium Oxide | PuO₂ | 276.06

### Conclusão:
A utilização de classes foi bastante útil, já que pode armazenar diversas instâncias, tanto na classe elemento, quanto na classe de moléculas. A partir desses conceitos, conseguimos criar um código bem consistente que consegue pegar um dataset de moléculas e conseguir obter a fórmula química e o peso molecular de qualquer molécula. Esta estrutura de objetos pode evoluir e armazenar outras informações associadas aos elementos químicos, servindo para futuras aplicações.

### Referências

- Documentação do pandas: PANDAS. pandas documentation — pandas 1.0.1 documentation. Disponível em: <https://pandas.pydata.org/docs/>.

- DataFrame da tabela periódica: PUBCHEM. PubChem. Disponível em: <https://pubchem.ncbi.nlm.nih.gov/>.